modules/fsi/src/interfacekernels/FluidStructureInterface.C

```diff
@@ -0,0 +1,152 @@
+{
+  Real r = 0;
+  switch (_component)
+  {
+    case 0:
+    {
+      switch (type)
+      {
+        case Moose::Element: // Element is fluid
+          r = _test[_i][_qp] * _D[_qp] * _neighbor_dotdot[_qp] * _normals[_qp](0);
+          break;
+
+        case Moose::Neighbor: // Neighbor is structure
+          r = _test_neighbor[_i][_qp] * -_u[_qp] * _normals[_qp](0);
+          break;
+      }
+      break;
+    }
+    case 1:
+    {
+      switch (type)
+      {
+        case Moose::Element: // Element is fluid
+          r = _test[_i][_qp] * _D[_qp] * _neighbor_dotdot[_qp] * _normals[_qp](1);
+          break;
+
+        case Moose::Neighbor: // Neighbor is structure
+          r = _test_neighbor[_i][_qp] * -_u[_qp] * _normals[_qp](1);
+          break;
+      }
+      break;
+    }
+    case 2:
+    {
+      switch (type)
+      {
+        case Moose::Element: // Element is fluid
+          r = _test[_i][_qp] * _D[_qp] * _neighbor_dotdot[_qp] * _normals[_qp](2);
+          break;
+
+        case Moose::Neighbor: // Neighbor is structure
+          r = _test_neighbor[_i][_qp] * -_u[_qp] * _normals[_qp](2);
+          break;
+      }
+      break;
+    }
+  }
+  return r;
+}
+
+Real
+FluidStructureInterface::computeQpJacobian(Moose::DGJacobianType type)
+{
+  Real jac = 0;
+  switch (_component)
+  {
+    case 0:
+    {
+      switch (type)
+      {
+        case Moose::ElementElement:
+          break;
+        case Moose::NeighborNeighbor:
+          break;
+        case Moose::ElementNeighbor:
+          jac = _test[_i][_qp] * _D[_qp] * _phi_neighbor[_j][_qp] * _neighbor_dotdot_du[_qp] *
+                _normals[_qp](_component);
+          break;
+        case Moose::NeighborElement:
+          jac = _test_neighbor[_i][_qp] * -_phi[_j][_qp] * _normals[_qp](_component);
+          break;
+      }
+      break;
+    }
+    case 1:
+    {
+      switch (type)
+      {
+        case Moose::ElementElement:
+          break;
+        case Moose::NeighborNeighbor:
+          break;
+        case Moose::ElementNeighbor:
+          jac = _test[_i][_qp] * _D[_qp] * _phi_neighbor[_j][_qp] * _neighbor_dotdot_du[_qp] *
+                _normals[_qp](_component);
+          break;
+        case Moose::NeighborElement:
+          jac = _test_neighbor[_i][_qp] * -_phi[_j][_qp] * _normals[_qp](_component);
+          break;
+      }
+      break;
+    }
+    case 2:
+    {
+      switch (type)
+      {
+        case Moose::ElementElement:
+          break;
+        case Moose::NeighborNeighbor:
+          break;
+        case Moose::ElementNeighbor:
+          jac = _test[_i][_qp] * _D[_qp] * _phi_neighbor[_j][_qp] * _neighbor_dotdot_du[_qp] *
+                _normals[_qp](_component);
+          break;
+        case Moose::NeighborElement:
+          jac = _test_neighbor[_i][_qp] * -_phi[_j][_qp] * _normals[_qp](_component);
+          break;
+      }
+      break;
+    }
+  }
+  return jac;
+}
```

Comment:
This code still need to be refactored per my previous suggestion

Reply:
Yes, we will incorporate your suggestions.

Reply:
@dschwen I have incorporated your suggestion. Let me know if further changes are required.